File: BSK1/src/bsk1/RailFence.java
package bsk1;

import java.util.ArrayList;

public class RailFence {
    
    public static String Encrypt(String plainText, int rail) {
        ArrayList<String> railFence = new ArrayList<String>();
        for (int i = 0; i < rail; i++) railFence.add("");
 
        int number = 0;
        int increment = 1;
        char[] arr = plainText.toCharArray();
    
        for(char c : arr)
        {
            if (number + increment == rail) increment = -1;
            else if (number + increment == -1) increment = 1;

            railFence.set(number, railFence.get(number)+c);
            number += increment;
        }
 
        String buffer = "";
        for(String s : railFence) buffer += s;
        return buffer;
    }

    public static String Decrypt(String cipherText, int rail) {
        int cipherLength = cipherText.length();
        ArrayList< ArrayList<Integer> > railFence = new ArrayList< ArrayList<Integer> >();
        for (int i = 0; i < rail; i++) 
            railFence.add(new ArrayList<Integer>());
 
        int number = 0;
        int increment = 1;
        for (int i = 0; i < cipherLength; i++)
        {
            if (number + increment == rail) increment = -1;
            else if (number + increment == -1) increment = 1;
            railFence.get(number).add(i);
            number += increment;
        }
 
        int counter = 0;
        char[] buffer = new char[cipherLength];
        for (int i = 0; i < rail; i++)
        {
            for (int j = 0; j < railFence.get(i).size(); j++)
            {
                buffer[ railFence.get(i).get(j) ] = cipherText.toCharArray()[counter];
                counter++;
            }
        }
 
        return new String(buffer);
    }
    
}

File: kryptografia/cryptography/railfence.py
# -*- coding: utf-8 -*-

__author__ = "Rafał Selewońko <rafal@selewonko.com>"


class RailFence(object):

    def encrypt(self, string, rail):
        """
        >>> rf = RailFence()
        >>> rf.encrypt("Ala ma kota", 3)
        'Amol akta a'
        >>>
        """
        railfence = ['' for x in range(rail)]
        number = 0
        increment = 1
        for char in string:
            if number + increment == rail:
                increment = -1
            elif number + increment == -1:
                increment = 1

            railfence[number] = railfence[number] + char
            number += increment

        return ''.join(railfence)

    def decrypt(self, string, rail):
        """
        >>> rf = RailFence()
        >>> rf.decrypt("Amol akta a", 3)
        'Ala ma kota'
        >>>
        """
        length = len(string)
        railfence = [[] for x in range(rail)]

        number = 0
        increment = 1
        for i in range(length):
            if number + increment == rail:
                increment = -1
            elif number + increment == -1:
                increment = 1
            railfence[number].append(i)
            number += increment

        counter = 0
        buffer = [None for x in range(length)]
        for i in range(rail):
            for j in range(len(railfence[i])):
                buffer[railfence[i][j]] = string[counter]
                counter += 1

        return ''.join(buffer)

if __name__ == "__main__":
    import doctest
    doctest.testmod()
